py/objtype: Add __dict__ attribute for class objects.

andrewleech opened the pull request:
The behavior mirrors instance object __dict__ attribute where a copy of the local attributes are provided.

The function is only enabled if `MICROPY_CPYTHON_COMPAT` is set, the same as the instance version of this function.

This can be useful to get access to the attributes on a class, eg. for https://github.com/micropython/micropython/issues/5322

jimmo commented:
Would it be worth making the returned dict readonly (setting is_fixed on the underlying map, just to make it really clear about thr behaviour?

andrewleech commented:
That does sound sensible. The same change would make sense on the instance __dict__ function too!

andrewleech commented:
@jimmo at the definition of `mp_map_t` the `is_fixed` flash comes with a comment of
```
// a fixed array that can't be modified; must also be ordered
```
https://github.com/micropython/micropython/blob/master/py/obj.h#L355

I'm not sure how firm that rule of needing to be ordered is, perhaps that flag should also be set in this case?

andrewleech commented:
@jimmo I've exposed `dict_copy` and used that, makes for much cleaner and safer implementation. 

I've not touched the `is_fixed` flag, unless we can be sure it's safe for use on dynamic dict objects I'm not comfortable using it.

andrewleech commented:
@dpgeorge I believe this PR is ready for review, it behaves well for me at least. A couple of others on slack have asked about this functionality recently too!

stinos commented:
> I've not touched the `is_fixed` flag, unless we can be sure it's safe for use on dynamic dict objects I'm not comfortable using it.

In any case a short comment in the code stating that a copy is returned meaning the original cannot be modified would be good, likewise for adding a test to the CPython difference documentation? And while we're at it maybe same for `__dict__` on instances since that has the same problem.

dpgeorge commented:
I think this is a good change, it mirrors the availability of this on instances.

But what I'd suggest to improve it is:
- wait for #6129 
- make the existing instance code use the new `mp_obj_dict_copy()` function to reduce code size (this might have different memory usage to inserting each element, but I think it'd be very similar for dicts that have not had any deletions, which is most likely the case for class/instance dicts)
- set `is_fixed=1` on the dict before returning it
- furthermore, as an optimisation, return `locals_dict` directly if it's already `is_fixed`

This should be a relatively small set of changes.

Also a test would eventually be needed, including a cpydiff test :)

andrewleech commented:
>     * make the existing instance code use the new `mp_obj_dict_copy()` function to reduce code size (this might have different memory usage to inserting each element, but I think it'd be very similar for dicts that have not had any deletions, which is most likely the case for class/instance dicts)
> 

I had originally thought to do this, however class `self->locals_dict` is a dict, so `mp_obj_dict_copy` just works whereas  instance `self->members` is a map so isn't conveniently compatible. 
Turns out wrapping it in a dict obj to pass it into dict_copy does still save almost 100bytes from flash though so is likely worth it
``` C
    if (attr == MP_QSTR___dict__) {
        // Create a new dict with a copy of the instance's map items.
        // This creates, unlike CPython, a 'read-only' __dict__: modifying
        // it will not result in modifications to the actual instance members.
        mp_obj_t members_dict = mp_obj_new_dict(0);
        mp_obj_dict_t *dict = MP_OBJ_TO_PTR(members_dict);
        dict->map = self->members;
        dest[0] = mp_obj_dict_copy(MP_OBJ_FROM_PTR(dict));
        return;
    }
```
from 
```
   text    data     bss     dec     hex filename
1841732     260  302044 2144036  20b724 firmware.elf
```
to
```
   text    data     bss     dec     hex filename
1841636     260  302044 2143940  20b6c4 firmware.elf
```
I also like your suggestions regarding the is_fixed flag too

dpgeorge commented:
> ```c
> mp_obj_t members_dict = mp_obj_new_dict(0);
> ```

No need to use heap memory, instead:
```c
mp_obj_dict_t dict;
dict.map = self->members;
dest[0] = mp_obj_dict_copy(MP_OBJ_FROM_PTR(&dict));
```

andrewleech commented:
'mp_obj_dict_copy' starts with `mp_check_self(mp_obj_is_dict_type(self_in));` so the bare struct wasn't going to be enough. I should be able to point the base object to something that'll satisfy the check though

dpgeorge commented:
> Also now with the next optimisation, in the case where the underlying instance map `is_fixed`, wont it need a real dict to return?

That optimisation won't work for instances, just classes.

andrewleech commented:
> That optimisation won't work for instances, just classes.

Thanks for the suggestions and clarification, I've cleaned that up now and added a basic unit test.

dpgeorge commented on py/objtype.c:
```diff
@@ -587,17 +587,16 @@ STATIC void mp_obj_instance_load_attr(mp_obj_t self_in, qstr attr, mp_obj_t *des
+        dict.base.type = &mp_type_dict;
+        dest[0] = mp_obj_dict_copy(MP_OBJ_FROM_PTR(&dict));
+        mp_obj_dict_t *dest_dict = MP_OBJ_TO_PTR(dest[0]);
+        dest_dict->map.is_fixed = 1;
```
Can you please put these changes to the instance `__dict__` lookup into a separate commit?

@stinos as the original author of this code, what are your thoughts on this change?  In particular do you think it's a good idea to set `is_fixed=1` to prevent modification of the returned dict (and so prevent subtle errors when the user tries to change the dict and expects those changes to be mirrored in the instance, which they are not)?

stinos replied:
Would be my preference to reduce chances of things being misused yes. And looking at the original discussion in https://github.com/micropython/micropython/pull/1757, I actually wanted it to be like that but that wasn't possible back then.

dpgeorge replied:
Ok, so let's keep the `is_fixed=1` part, to make the returned dict R/O.  It will be a breaking change though, if users are relying on the dict being modifiable (eg because they know it's not reflected in the instance but they want anyway to reuse the dict for something else).  But it'll fail pretty quickly now if someone is doing this and they can fix their code by simply doing `d = dict(my_type.__dict__)`

dpgeorge commented on tests/basics/class_dict.py:
```diff
@@ -0,0 +1,10 @@
+
+class Foo:
+    self.a = 1
+    self.b = "bar"
```
`self` shouldn't be needed here

andrewleech replied:
whoops, fixed along with splitting the commits.

pi-anl committed:
py/objtype: Add __dict__ attribute for class objects.

pi-anl committed:
py/objtype: Use mp_obj_dict_copy() for creating obj.__dict__ attribute.

dpgeorge commented on py/objtype.c:
```diff
@@ -1013,6 +1012,21 @@ STATIC void type_attr(mp_obj_t self_in, qstr attr, mp_obj_t *dest) {
+        if (attr == MP_QSTR___dict__) {
+            // Returns a read-only dict of the class attributes.
+            // If the internal locals is not fixed, a copy will be created.
+            mp_obj_dict_t *dict = MP_OBJ_FROM_PTR(self->locals_dict);
```
doesn't need a cast

dpgeorge commented on py/objtype.c:
```diff
@@ -1013,6 +1012,21 @@ STATIC void type_attr(mp_obj_t self_in, qstr attr, mp_obj_t *dest) {
+            // If the internal locals is not fixed, a copy will be created.
+            mp_obj_dict_t *dict = MP_OBJ_FROM_PTR(self->locals_dict);
+            if (dict->map.is_fixed) {
+                dest[0] = dict;
```
needs `MP_OBJ_FROM_PTR`

dpgeorge commented on py/objtype.c:
```diff
@@ -1013,6 +1012,21 @@ STATIC void type_attr(mp_obj_t self_in, qstr attr, mp_obj_t *dest) {
+            if (dict->map.is_fixed) {
+                dest[0] = dict;
+            } else {
+                dest[0] = mp_obj_dict_copy(dict);
```
needs `MP_OBJ_FROM_PTR`
